gobblin-tunnel/src/main/java/gobblin/tunnel/Tunnel.java

```
@@ -0,0 +1,178 @@
+ *    the target host via the intermediate proxy.)
+ * 7. When the Gobblin extractor (e.g. JDBC data source) is closed down, the Tunnel must be shut down as well.
+ *
+ * The Tunnel can accept as many connections as the JdbcExtractor opens. It uses NIO to minimize resource usage.
+ *
+ * @author navteniev@linkedin.com
+ * @author kkandekar@linkedin.com
+ */
+public class Tunnel {
+  public static final int NON_EXISTENT_PORT = -1;
+
+  private static final Logger LOG = LoggerFactory.getLogger(Tunnel.class);
+
+  private ServerSocketChannel server;
+  private Thread thread;
+  private final Config config;
+
+  private Tunnel(String remoteHost, int remotePort, String proxyHost, int proxyPort) {
+    config = new Config(remoteHost, remotePort, proxyHost, proxyPort);
+  }
+
+  private Tunnel open() throws IOException {
+    try {
+      server = ServerSocketChannel.open().bind(null);
+      startTunnelThread();
+
+      return this;
+    } catch (IOException ioe) {
+      LOG.error("Failed to open the tunnel", ioe);
+      throw ioe;
+    }
+  }
+
+  public int getPort() throws IOException {
+    SocketAddress localAddress = null;
+
+    try {
+      if (server != null && server.isOpen()) {
+        localAddress = server.getLocalAddress();
+      }
+      if (localAddress instanceof InetSocketAddress) {
+        return ((InetSocketAddress) localAddress).getPort();
+      }
+    } catch (IOException e) {
+      LOG.error("Failed to get tunnel port", e);
+      throw e;
+    }
+    return NON_EXISTENT_PORT;
+  }
+
+  private void startTunnelThread() {
+    thread = new Thread(new Dispatcher(), "Tunnel Listener");
+    thread.start();
+  }
+
+  public boolean isTunnelThreadAlive() {
+    return (thread != null && thread.isAlive());
+  }
+
+  private class Dispatcher implements Runnable {
+
+    private Selector selector;
+
+    public Dispatcher() {
+      try {
+        selector = Selector.open();
```

Good idea. This will also allow the open() method to fail early and cleanly if Selector#open fails.